### PR TITLE
feat(fetch): maxRetries for fetch

### LINKED ISSUE #30978: [Feature]: retry-mechanism for API request incase failure due to ECONNRESET
### 🚀 Feature Request

ECONNRESET might happen due to connectivity issues. As multiple executions are happening through API calls, sometimes randomly playwright through ECONNRESET error.  I used my logic of retrying using try-catch apiContext.post/get, the error object is undefined on ECONNRESET occurs. It would be nice playwright threw the error.statuscode as  ECONNRESET / provided an option to retry the API

    let retryCount = 0;
    let maxRetries = 3;
    while (retryCount < maxRetries) {
      try {
        const Response = await apiContext.get(URL, {
          headers: {
            'Connection': 'keep-alive',
            'Authorization': SessionToken,
            'Content-Type': 'application/json',
          },
          timeout: 0,
          failOnStatusCode: true
        });
        await Response.json();
      }
     } catch (error) {
        console.log(error); 
        if (error.code.toString().includes("ECONNRESET")) {
          retryCount++;
          console.log(`Connection reset. Retrying... (${retryCount}/${maxRetries})`);
          continue;
        } else {
          throw error;
        }
      }

### Example

_No response_

### Motivation

This feature will help API Automation better,reliable & saves the time of retrigger of entire automation suite

### PR DESCRIPTION
**Pull request by yury-s**
Fixes https://github.com/microsoft/playwright/issues/30978

### CODE REVIEW

File: docs/src/api/params.md [hunk 1]
```diff
@@ -458,6 +458,12 @@ Whether to ignore HTTPS errors when sending network requests. Defaults to `false
 Maximum number of request redirects that will be followed automatically. An error will be thrown if the number is exceeded.
 Defaults to `20`. Pass `0` to not follow redirects.
 
+## js-python-csharp-fetch-option-maxretries
+* langs: js, python, csharp
+- `maxRetries` <[int]>
+
+Maximum number of times socket errors should be retried. Currently only `ECONNRESET` error are retried. An error will be thrown if the limit is exceeded. Defaults to `5`. Pass `0` to disable retries.
```
Comment (dgozman) [inline docs/src/api/params.md:465]:
```suggestion
Maximum number of times socket errors should be retried. Currently only `ECONNRESET` error is retried. An error will be thrown if the limit is exceeded. Defaults to `5`. Pass `0` to disable retries.
```

File: packages/playwright-core/src/server/fetch.ts [hunk 2]
```diff
@@ -247,6 +247,26 @@ export abstract class APIRequestContext extends SdkObject {
...
   }
 
+  private async _sendRequestWithRetries(progress: Progress, url: URL, options: SendRequestOptions, postData?: Buffer): Promise<Omit<channels.APIResponse, 'fetchUid'> & { body: Buffer }>{
+    const maxRetries = 5;
+    let backoff = 250;
+    for (let i = 0; i <= maxRetries; i++) {
+      try {
+        return await this._sendRequest(progress, url, options, postData);
+      } catch (e) {
+        if (i === maxRetries - 1 || (options.deadline && monotonicTime() + backoff > options.deadline))
```
Comment (dgozman) [inline packages/playwright-core/src/server/fetch.ts:257]:
```suggestion
        if (i === maxRetries || (options.deadline && monotonicTime() + backoff > options.deadline))
```

File: packages/playwright-core/src/server/fetch.ts [hunk 3]
```diff
@@ -247,6 +247,26 @@ export abstract class APIRequestContext extends SdkObject {
...
+          throw new Error(`Failed after ${i + 1} attempt(s): ${e}`);
+        // Retry on connection reset only.
+        if (e.code !== 'ECONNRESET')
+          throw e;
+        progress.log(`Received ECONNRESET, will retry after ${backoff}ms.`);
```
Comment (dgozman) [inline packages/playwright-core/src/server/fetch.ts:262]:
```suggestion
        progress.log(`  Received ECONNRESET, will retry after ${backoff}ms.`);
```

File: packages/playwright-core/src/server/fetch.ts [hunk 4]
```diff
@@ -247,6 +247,26 @@ export abstract class APIRequestContext extends SdkObject {
     }
   }
 
+  private async _sendRequestWithRetries(progress: Progress, url: URL, options: SendRequestOptions, postData?: Buffer): Promise<Omit<channels.APIResponse, 'fetchUid'> & { body: Buffer }>{
+    const maxRetries = 5;
```
Comment (dgozman) [inline packages/playwright-core/src/server/fetch.ts:251]:
```suggestion
    const maxRetries = options.maxRetries ?? 5;
```

File: docs/src/api/params.md [hunk 5]
```diff
@@ -458,6 +458,12 @@ Whether to ignore HTTPS errors when sending network requests. Defaults to `false
 Maximum number of request redirects that will be followed automatically. An error will be thrown if the number is exceeded.
 Defaults to `20`. Pass `0` to not follow redirects.
 
+## js-python-csharp-fetch-option-maxretries
```
Comment (dgozman) [inline docs/src/api/params.md:461]: What about Java?

**Review by dgozman — APPROVED** [review]

File: docs/src/api/params.md (hunk 5 above)
Comment (yury-s) [reply docs/src/api/params.md:461]: The option is in RequestOptions

File: docs/src/api/params.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (yury-s) [reply docs/src/api/params.md:465]: In that case, we should just make it `0' by default and let the users opt-in. This way they'll have to think about the implications. WDYT?

[the remark this answers is not in the document]
Comment (yury-s) [reply docs/src/api/params.md]: Change to 0 by default, will discuss it during 1.46 API review.

**Pull request merged**